src/field_theory/polynomial_galois_group.lean

```diff
@@ -0,0 +1,245 @@
+/-
+Copyright (c) 2020 Thomas Browning and Patrick Lutz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Thomas Browning and Patrick Lutz
+-/
+
+import field_theory.galois
+
+/-!
+# Galois Groups of Polynomials
+
+In this file we introduce the Galois group of a polynomial, defined as
+the automorphism group of the splitting field.
+
+## Main definitions
+
+- `gal p`: the Galois group of a polynomial p.
+- `restrict p E`: the restriction homomorphism `(E ≃ₐ[F] E) → gal p`.
+- `gal_action p E`: the action of `gal p` on the roots of `p` in `E`.
+
+## Main results
+
+- `restrict_smul`: `restrict p E` is compatible with `gal_action p E`.
+- `gal_action_hom_injective`: the action of `gal p` on the roots of `p` in `E` is faithful.
+- `restrict_prod_inj`: `gal (p * q)` embeds as a subgroup of `gal p × gal q`.
+-/
+
+noncomputable theory
+open_locale classical
+
+open finite_dimensional
+
+namespace polynomial
+
+variables {F : Type*} [field F] (p q : polynomial F) (E : Type*) [field E] [algebra F E]
+
+/-- The Galois group of a polynomial -/
+@[derive [has_coe_to_fun, group, fintype]] def gal := p.splitting_field ≃ₐ[F] p.splitting_field
+
+namespace gal
+
+instance [h : fact (p.splits (ring_hom.id F))] : unique p.gal :=
+{ default := 1,
+  uniq := λ f, alg_equiv.ext (λ x, by { obtain ⟨y, rfl⟩ := algebra.mem_bot.mp
+    ((subalgebra.ext_iff.mp ((is_splitting_field.splits_iff _ p).mp h) x).mp algebra.mem_top),
+    rw [alg_equiv.commutes, alg_equiv.commutes] }) }
+
+instance : unique (0 : polynomial F).gal :=
+begin
+  haveI : fact ((0 : polynomial F).splits (ring_hom.id F)) := splits_zero _,
+  apply_instance,
+end
+
+instance [h : fact (p.splits (algebra_map F E))] : algebra p.splitting_field E :=
+(is_splitting_field.lift p.splitting_field p h).to_ring_hom.to_algebra
+
+instance [h : fact (p.splits (algebra_map F E))] : is_scalar_tower F p.splitting_field E :=
+is_scalar_tower.of_algebra_map_eq
+  (λ x, ((is_splitting_field.lift p.splitting_field p h).commutes x).symm)
+
+/-- The restriction homomorphism -/
+def restrict [h : fact (p.splits (algebra_map F E))] : (E ≃ₐ[F] E) →* p.gal :=
+alg_equiv.restrict_normal_hom p.splitting_field
+
+section roots_action
+
+/-- The roots of `p` in `E` -/
+def roots : set E := (p.map (algebra_map F E)).roots.to_finset
+
+/-- The roots of `p` in `p.splitting_field` -/
+def roots_aux := roots p p.splitting_field
```

Comment:
I'm not totally sold on the name `gal.roots_aux`, since this feels like it would be useful outside of the Galois group itself. How about `splitting_field.roots`? (Although I guess that that would cause an import cycle, so if it's too much work just keep it here. I'm thinking about refactoring `polynomial.roots` anyway to take an `algebra` or `ring_hom`, like `splits` does.)

Reply:
I feel like `splitting_field.roots` would mean the multiset, rather than the set. If you think that these definitions will be used outside of the file, should they be named something other than `roots`, seeing as `roots` is already used to refer to the multiset?

Reply:
That's true, how about `roots : multiset` from `root_set : set`?